receiver/dynamic/receiver.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamic
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/open-telemetry/opentelemetry-collector/config"
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+)
+
+var (
+	errNilNextConsumer = errors.New("nil nextConsumer")
+	factories          = buildFactoryMap(&config.ExampleReceiverFactory{})
+)
+
+type factoryMap map[string]receiver.Factory
+
+func (fm factoryMap) get(receiverType string) (receiver.Factory, error) {
+	if factory, ok := factories[receiverType]; ok {
+		return factory, nil
+	}
+	return nil, fmt.Errorf("factory does not exist for receiver type %q", receiverType)
+}
+
+func buildFactoryMap(factories ...receiver.Factory) factoryMap {
+	ret, err := receiver.Build(factories...)
+	if err != nil {
+		panic(fmt.Sprintf("failed to build factory map for dynamic receiver: %s", err))
+	}
+	return ret
+}
+
+var _ receiver.MetricsReceiver = (*dynamicReceiver)(nil)
+
+// dynamicReceiver implements consumer.MetricsConsumer.
+type dynamicReceiver struct {
+	nextConsumer consumer.MetricsConsumer
+	logger       *zap.Logger
+	cfg          *Config
+	receivers    []receiver.Receiver
+}
+
+// New creates the dynamic receiver with the given parameters.
+func New(logger *zap.Logger, nextConsumer consumer.MetricsConsumer, cfg *Config) (receiver.MetricsReceiver, error) {
+	if nextConsumer == nil {
+		return nil, errNilNextConsumer
+	}
+
+	r := &dynamicReceiver{
+		logger:       logger,
+		nextConsumer: nextConsumer,
+		cfg:          cfg,
+	}
+	return r, nil
+}
+
+// loadRuntimeReceiverConfig loads the given subreceiverConfig merged with config values
+// that may have been discovered at runtime.
+func (dr *dynamicReceiver) loadRuntimeReceiverConfig(subConfig *subreceiverConfig, config map[string]interface{}) (configmodels.Receiver, error) {
```

Comment:
So just to clarify. `subConfig` contains what is statically specified in the config file, `config` contains settings that may be applied at runtime and will override anything that already exists in `subConfig`. Do I understand it correctly?

Reply:
Correct.

Reply:
Would it make sense to rename parameters to reflect this? Perhaps:
```suggestion
func (dr *dynamicReceiver) loadRuntimeReceiverConfig(userDefinedSubConfig *subreceiverConfig, dynamicSubConfig map[string]interface{}) (configmodels.Receiver, error) {
```

Reply:
Yeah I don't think the names here are great.

Would subConfigFromFile and subConfigFrom{Env/Observer} make it clear where it's coming from as as userDefinedSubConfig is still a bit ambiguous. (Is it user defined from the config file or user defined from the annotations?)

Reply:
Technically the user-defined config does not necessarily come from a file, since we very recently added ability to supply the config programmatically without having an actual config file.

But I agree "user-defined" is ambiguous.

Perhaps "staticSubConfig" and "subConfigFromEnv/Observer" is a bit better?